internal/pkg/vcsutil/vcs_util_test.go

```diff
@@ -0,0 +1,362 @@
+package vcsutil
+
+import (
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"testing"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/stretchr/testify/require"
+)
+
+type VCSTester struct {
+	repoPath       string
+	testFiles      []*os.File
+	remoteRepoPath string
+	remoteUrl      string
+	remoteName     string
+}
+
+var testHasGit bool
+
+func init() {
+	if _, err := exec.LookPath("git"); err == nil {
+		testHasGit = true
+	}
+}
+
+func generateGitState(branchName string, t *testing.T) (VCSTester, error) {
+
+	log := hclog.Default()
+
+	// Create a temporary directory for our test
+	td := t.TempDir()
+
+	// Create a temporary directory for our remote
+	remote := t.TempDir()
+
+	if _, err := runGitCommand(log, td, "init", "-b", branchName); err != nil {
+		return VCSTester{}, err
+	}
+
+	if _, err := runGitCommand(log, remote, "init"); err != nil {
+		return VCSTester{}, err
+	}
+
+	remoteUrl := remote + "/.git"
+
+	remoteName := "remote"
+	// add remote
+	if _, err := runGitCommand(log, td, "remote", "add", remoteName, remoteUrl); err != nil {
+		return VCSTester{}, err
+	}
+
+	// Create a test file and commit
+	file := td + "/testfile"
+	r, err := os.OpenFile(file, os.O_CREATE, 0600)
+	if err != nil {
+		return VCSTester{}, err
+	}
+	r.Close()
+
+	if _, err := runGitCommand(log, td, "add", "testfile"); err != nil {
+		return VCSTester{}, err
+	}
+
+	// Add another one for testing multi-line diff output
+	f := td + "/second-testfile"
+	rr, err := os.OpenFile(f, os.O_CREATE, 0600)
+	if err != nil {
+		return VCSTester{}, err
+	}
+	rr.Close()
+
+	if _, err := runGitCommand(log, td, "add", "second-testfile"); err != nil {
+		return VCSTester{}, err
+	}
+
+	if _, err := runGitCommand(log, td, "commit", "-m", "'first commit'"); err != nil {
+		return VCSTester{}, err
+	}
+
+	if _, err := runGitCommand(log, td, "push", remoteName, branchName); err != nil {
+		return VCSTester{}, err
+	}
+
+	return VCSTester{
+		td,
+		[]*os.File{r, rr},
+		remote,
+		remoteUrl,
+		remoteName,
+	}, nil
+}
+
+func TestIsDirty(t *testing.T) {
+	if !testHasGit {
+		t.Skip("git not installed")
+		return
+	}
+
+	hclog.Default().SetLevel(hclog.Debug)
+
+	require := require.New(t)
+	branchName := "waypoint"
+
+	vcsTester, err := generateGitState(branchName, t)
+	require.NoError(err)
+
+	v, err := NewVcsChecker(
+		hclog.Default(),
+		vcsTester.repoPath,
+		vcsTester.remoteUrl,
+	)
+
+	require.NoError(err)
+
+	t.Run("Initial state is clean", func(t *testing.T) {
+		dirty, err := v.RepoHasDiff(branchName)
+		require.NoError(err)
+		require.False(dirty)
+	})
+
+	t.Run("Creating (but not committing) a new file is dirty", func(t *testing.T) {
```

Comment:
Given the above note (on `DOTgit`), you can then just make a series of exact-state git repo folders rather than dynamically recreating the state in your test.